design/rest-api-guideline.md

```diff
@@ -0,0 +1,434 @@
+
+```
+HTTP/1.1 204 No Content
+Content-Length: 0
+```
+
+#### PATCH
+
+HTTP `PATCH` should be used to partially update the resource state.
+
+Successful `PATCH` requests should always result in a `204` status code and empty response.
+
+```
+HTTP/1.1 204 No Content
+Content-Length: 0
+```
+
+Partial updates should follow the [RFC7396 - Json Merge Patch](https://tools.ietf.org/html/rfc7396) standard. 
+
+The `PATH` operation should be driven by a specific `Content-Type` header as defined by RFC7396 and other patch strategies. By using the
+`Content-Type` as a selector, this guideline can support additional patch strategies in the future.
+
+##### Note on RFC7396 - Json Merge Patch
+
+The RFC7396 is the most simple `PATH` strategy, but it has some limitations. The limitations are basically around updating lists on a resource which will be entirely replaced by the patch.
+
+Under certain circumstances, other patch strategies may be useful thus the constraint around the `Content-Type` header of `PATH` requests.
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `PUT` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|204|
+|400|
+|401|
+|403|
+|404|
+|409|
+|415|
+|500|
+
+#### DELETE
+
+HTTP `DELETE` should be used to remove a resource.
+
+Successful `DELETE` requests should always result in a `204` status code and empty response.
+
+```
+HTTP/1.1 204 No Content
+Content-Length: 0
+```
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `DELETE` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|204|
+|400|
+|401|
+|403|
+|404|
+|500|
+
+## Practices
+
+### Error Handling
+
+A good error handling mechanism helps to communicate to clients the reasons why a request could not be fulfilled and eventually
+recover from this errors.
+
+When returning errors, developers should use a common [HTTP Status Code](#http-status-codes) and include a body using JSON as the format with the information
+why a request failed.
+
+The error message should have the following format:
+
+|Claim|Description|Required|
+|---|---|---|
+|error|The error code|Yes|
+|error_description|A text providing more details for clients about the error and eventually how to recover from it|No|
+
+Here is an example of a error response:
+
+```json
+HTTP/1.1 400 Bad Request
+{
+    "error": "invalid_request",
+    "error_description": "The field 'email' is mandatory"
+}
+```
+
+#### Error Codes
+
+Error codes can be any string. They should be lowercase and use `_` when the code has multiple words. However, developers should 
+consider existing error codes before creating new ones.
+
+When processing error responses, developers should look the error codes being used by other APIs and see if they can be reused.
+
+### Filtering
+
+TODO
+
+### Pagination
+
```

Comment:
@pedroigor I don't know what you think about the proposal below. But let me know if makes sense:

### Pagination

For large sets of data, since the early releases pagination and filtering capabilities were [provided on Keycloak](https://www.keycloak.org/docs-api/6.0/rest-api/index.html) by using a cursor (*first*) and page size (*max*).

For example, to retrieve a list of the first 10 groups starting at the first item we provide:

* Request

```
curl -I "http://localhost:8081/auth/admin/realms/master/groups?first=0&max=10" -H "Authorization: bearer $TKN" -H "Accept: application/json" -H "Authorization: Bearer $TKN"
```
* Response headers

```
HTTP/1.1 200 OK
Connection: keep-alive
Cache-Control: no-cache
Content-Type: application/json
Content-Length: 961
Date: Wed, 22 May 2019 13:44:57 GMT
```

#### Possible enhancements

Today, the response headers do not provide too much detail about the number of items we have into the database. Adding that would be handy for the new account console and also other people [integrating with our APIs](http://lists.jboss.org/pipermail/keycloak-user/2019-May/018108.html).

Please, notice that this is just an enhancement to the current implementation approach as we have the opportunity to design something new. For example:

* Request

```
curl -I "http://localhost:8081/auth/admin/realms/master/groups" -H "Authorization: bearer $TKN" -H "Accept: application/json" -H "Authorization: Bearer $TKN"
```
* Response headers

```
HTTP/1.1 200 OK
Connection: keep-alive
Cache-Control: no-cache
Content-Type: application/json
Content-Length: 961
Date: Wed, 22 May 2019 13:44:57 GMT

Link: <http://localhost:8081/auth/admin/realms/master/groups&page=2>; rel="next",
  <http://localhost:8081/auth/admin/realms/master/groups&page=40>; rel="last"
```

[GitHub uses this approach](https://developer.github.com/v3/guides/traversing-with-pagination/) through [RFC 5988 for Web Linking](https://tools.ietf.org/html/rfc5988), which is mentioned in the HATEOAS section from the [REST API Tutorial](https://restfulapi.net/hateoas/). Web Linking is something that already comes with JAX-RS API.

Alternatively, people should be able to query the number of items on a page.

Reply:
The approach looks correct and aligned with some common practices we see out there.

I'm not sure if we really need to calculate the total count of entries and what exactly is the last page. Today our pagination have options for `First`, `Previous` and `Next Page`. We basically use the {{first}} and {{max}} parameters which I think is enough.

However, there is a behavior that we could certainly improve. Today, if you try to click the `Next Page` button you may get a empty result set. I would suggest that we perform queries based on N + 1 where N is the number of items on a page and use something to indicate the client that there is no next page. Using a {{Link}} header as you suggested may do the trick.

Reply:
> The approach looks correct and aligned with some common practices we see out there.
> 
> I'm not sure if we really need to calculate the total count of entries and what exactly is the last page. Today our pagination have options for `First`, `Previous` and `Next Page`. We basically use the {{first}} and {{max}} parameters which I think is enough.

@pedroigor I just updated the idea. But in summary we're not going to calculate the total of entries. But let consumers know which one is the last page, so they have an idea about how many items are available to consume. This can be useful for people, because they don't need to iterate over all the entries to know the amount of data they gonna consume. This is what happens today, people iterate over it until they hit an empty response.

Alternatively people could provide also provide `max` as we already do today. I can update my comments if you think it makes sense.

> 
> However, there is a behavior that we could certainly improve. Today, if you try to click the `Next Page` button you may get a empty result set. I would suggest that we perform queries based on N + 1 where N is the number of items on a page and use something to indicate the client that there is no next page. Using a `Link` header as you suggested may do the trick.

I'm not sure if I follow your idea. But let me try to elaborate how it was supposed to work, maybe we are talking about the same things. `rel="next"` will always return N + 1, so let's assume a scenario where the last page is 50 and I'm on page 49. That means that when I hit the page 49 I should get:

Link: <http://localhost:8081/auth/admin/realms/master/groups&page=50>; rel="next",
  <http://localhost:8081/auth/admin/realms/master/groups&page=50>; rel="last"

So if we hit next, the interface should be aware that page 50 is the last based on the response headers.

Does make any sense to you?

Reply:
In order to return `rel=last` you need to calculate the last page somehow. I don't understand how you are going to do that given that you also said we are not going to calculate the total of entries.

My last comment is about returning *only* `rel=next` to indicate to the client whether or not there is a next page. If the response contains the header the client can then render the `Next` button given that the server is indicating that there is more pages. Otherwise, the client can disable/hide the button as it reached the last page.

Considering your scenario where the client asks for page 49. Suppose `max` is 20. The server will query for 21 entries (starting on `first` which represents the actual page). If the query returns 21 it means we have a `rel=next`. Otherwise, there are no more pages and you don't return `rel=nex`.

Reply:
> In order to return `rel=last` you need to calculate the last page somehow. I don't understand how you are going to do that given that you also said we are not going to calculate the total of entries.

I got your point now, I thought your comment was related with my last sentence where the idea of returning the total. And of course there are performance penalties when we calculate the total of records just to display the last page.

For now, let's remove it and if more people request we can always iterate over it.

I just updated it, let me know if that works for you.

> 
> My last comment is about returning _only_ `rel=next` to indicate to the client whether or not there is a next page. If the response contains the header the client can then render the `Next` button given that the server is indicating that there is more pages. Otherwise, the client can disable/hide the button as it reached the last page.
> 
> Considering your scenario where the client asks for page 49. Suppose `max` is 20. The server will query for 21 entries (starting on `first` which represents the actual page). If the query returns 21 it means we have a `rel=next`. Otherwise, there are no more pages and you don't return `rel=nex`.

Reply:
@abstractj, updated the PR with what we discussed.

Reply:
@pedroigor thank you. Marking it as resolved.